walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
 
 package com.google.sps;
 
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
+import com.google.sps.MeetingRequest;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, 
+                                     MeetingRequest request) {
+    // Get unavailable time for all required attendees in meeting request
+    ArrayList<TimeRange> unavailable = new ArrayList<>();
+    for (String attendee : request.getAttendees()) {
+      ArrayList<TimeRange> meetingTimes = getMeetingTimes(events, attendee);
+      merge(unavailable, meetingTimes);
+    }
+
+    // Get time ranges that are longer than the requested meeting duration
+    Collection<TimeRange> result = getAvailability(unavailable);
+    result.removeIf( 
+      timeRange -> timeRange.duration() < request.getDuration());
+
+    return result;
+  }
+
+  /**
+   * Adds the unavailable time ranges of a new attendee to the cumulative
+   * unavailable time ranges.
+   * Resolves conflicts by merging overlapping ranges into new ranges.
+   *
+   * @param result Cumulative unavailable time ranges
+   * @param other Unavailable time ranges of a new attendee
+   */
+  private void merge(ArrayList<TimeRange> result, ArrayList<TimeRange> other) {
+    if (result.isEmpty()) {
+      result.addAll(other);
+      return;
+    }
+    if (other.isEmpty()) {
+      return;
+    }
+
+    result.addAll(other);
+    Collections.sort(result, TimeRange.ORDER_BY_START);
+
+    // Merge overlapping ranges
+    for (int i = 0; i < result.size()-1; i++) {
+      TimeRange current = result.get(i);
+      TimeRange next = result.get(i+1);
+
+      if (current.overlaps(next)) {
+        if (current.contains(next)) {
+          // Case 1:  |_______|
+          //            |__|
+          result.remove(next);
+        } else {
+          // Case 2:  |______|
+          //               |______|
+          result.set(i, TimeRange.fromStartEnd(current.start(), 
+                                               next.end(), false));
+          result.remove(next);
+          i--;
```

Comment:
Got it!
I realize that for Case 1, i-- is also needed.
To deal with scenario like
<img width="137" alt="Screen Shot 2020-07-01 at 7 39 20 PM" src="https://user-images.githubusercontent.com/37273127/86239814-970d9f00-bbd2-11ea-901b-0c9ea214a57c.png">

Reply:
emmm I would actually suggest to sort these segments with first key start and second key end.
So we don't need to deal with these corner cases, which is hard to catch.

Reply:
I'm not sure how to use two keys to avoid the corner cases, could you specify on this? thank you!

Reply:
Oh, never mind.

A simpler way might be
```
for (int i = 0; i + 1 < result.size(); ) {
  TimeRange current = result.get(i), next = result.get(i+1);
  if(current.overlaps(next)) {
    result.set(i, TimeRange.fromStartEnd(
        current.start(), max(current.end(), next.end()), false));
    // Keep checking the current one.
    result.remove(i+1);
  } else {
    i++;
  }
}
```

Reply:
PS: Use your own judgement.

Reply:
Got it, thank you!